### ISSUE TITLE
[invokers] - how to style invoking element based on Invoker state?

### ISSUE THREAD

**Issue by scottaohara** [openui/open-ui#622]
As per the discussion on https://github.com/openui/open-ui/issues/610 we briefly discussed how there is not presently a way to style the invoking element of a popup based on the current state.

Presently, if building something like this with HTML / ARIA today, one could use `<button aria-expanded=true|false>` and use the ARIA attribute as the hook.  It seems this is something the invoking element for a popup should/could support (whether an attribute needs to be reflected in the DOM, or a pseudo class (`:expanded` and `:collapsed` to match current ARIA conventions?  `:opened` and `:closed`), or other e.g., `.indeterminate` for checkboxes in HTML.  Not implying preference on any of those, merely listing the ways to do this that quickly come to mind.

cc @mfreed7

**Comment by hidde** [comment]
As we resolved on applying `:open` to `<selectmenu>` (to which `:close` was added) in https://github.com/openui/open-ui/issues/547… would it make sense (and technically possible?) to have `:open` apply to all popup invoking elements?

**Comment by jh3y** [comment]
Jus' throwing some extra things in here for thought:

- Although not the prettiest you can do this with `:has` but that does open up a maintenance issue.
```css
body:has([id=pop]:open) [popover=pop] {
  /* Open styles */
}
```
- Are there many examples out there currently of popovers that have a toggle button aside from `selectmenu` or something involving a menu like interaction?

**Comment by scottaohara** [comment]
Sub-navigations, popup disclosure widgets / toggle tips that may contain info, links, buttons. 

those should be rather common in Regards to needing popup behavior.

**Comment by jh3y** [comment]
Then, if some sort of `open` attribute could be used, I'm imagining it would be something like `popoveropen` based on other attribute names, this might in turn solve #631. The existence of the attribute could potentially drive the show/hide behavior and means we don't get something like 

```html
<div popup popupopen popupdefaultopen>
```

That reminds me of React when I'm having to define `defaultValue` 😅

**Comment by jh3y** [comment]
There's maybe a more dynamic way:

```html
<div class=toggle-tip-container toggle-container>
  <button popovertoggletarget=pop>
    Toggle
  </button>
  <div popover id=pop>Popover</div>
</div>
```

```css
[toggle-container]:has([popover]:open) button {
  /* Do something to my button! */
}
```

**Comment by mfreed7** [comment]
So I notice the action item in the notes above for me. I'm a bit unclear on the exact mechanics of the proposal, and in fact I heard two variations, one from @dbaron and one from @gregwhitworth. I just tried to write them both up, and I realized I'm not sure I completely understand them. Both were essentially (functional?) pseudo elements that could be used to associate a popover invoker with its popover, and also perhaps a label with its form control. In the popover case, the idea is to be able to style the button based on the `:open` state of its associated popover. In the form control case, style the `<label>` based on the `:hover` state of its form control.

@dbaron and @gregwhitworth can you please fill in the gaps above?

**Comment by gregwhitworth** [comment]
What mine was that in EdgeHTML we propogated state to the associated element. Using the historical example:

```
HTML
===========
<label for=foo>My Label</label>
<button id=foo>Hello</button>

CSS
===========
label:hover {
    background: blue;
}
```

In order to achieve this today, authors wrap the `button` with the `label` which is akin to @jh3y CSS approach using `has`.

So I was merely stating that we should define the anchor attributes to propagate style psuedo class state to the button that has an association.

@dbaron was referring to a complex constructor or function that would allow association to be called. I'd need to see the definition of this to ensure that dynamic solutions can be covered.

**Comment by mfreed7** [comment]
I like this idea of allowing easy(er than `:has`) styling of a triggering element, based on the state of the triggered popover. However, I think a) it might be a more general feature (see above), and b) I think it can almost surely be added later without breaking anything. For these two reasons, I'm going to mark this as a "V2" feature. Hopefully ok with folks.

**Comment by lukewarlow** [comment]
Gonna commandeer this issue to discuss styling invokers based on Invokee state more generally than just popover.

**Comment by lukewarlow** [comment]
https://github.com/w3c/csswg-drafts/issues/9547 contains some discussion regarding this.

**Comment by scottaohara** [comment]
it is probably a good idea to have a more generic invoked (true) / not invoked (false) styling hook, rather than try to tap into any explicit state.  Especially since the expanded/collapsed state is only applicable to some, and i still have doubts about whether automatically exposing a pressed/not pressed state for play/pause makes sense - since people are more likely to change the labels of those, and we don't want a label AND state change to occur.

**Comment by bkardell** [comment]
> it is probably a good idea to have a more generic invoked (true) / not invoked (false) styling hook

Is it invocation related or state related? Like, imagine we had a silly interface:

```html
<button id="A" invoketarget="my-video" invokeaction="playpause">Play/Pause</button>

<video id="my-video"></video>

<button  id="B" invoketarget="my-video" invokeaction="playpause">Play/Pause</button>
```

* If the video autoplayed or something: Do either of the invokers light up?
* Which is "invoked" here - play, or pause - or does it depend on whether the video was already playing?
* If you click button A, and we determine it should match - does B also match?

**Comment by scottaohara** [comment]
i didn't anticipate this working (well) without being state related.  maybe that's a bad assumption on my part, but if the video was autoplaying i would have anticipated both those buttons reflecting the 'playing' state of the video - so 'invoked = true' for the purposes of stying (yes i know it's not invoked by the user - but also the alternative is rubbish UX if not reflecting the state of the video).  activating either button would cause the video to pause, and both buttons would need to refelct this state and expose an invoked false style hook.

**Comment by keithamus** [comment]
I guess today, one might use `:has()` to solve this problem for simple cases:

```html
<style>
  body:has(video:playing) button .play-label,
  body:has(video:paused) button .pause-label {
    display:none
  }
</style>
<button id="A" invoketarget="my-video" invokeaction="playpause">
  <span class="play-label">Play</span>
  <span class="pause-label">Pause</span>
</button>
<video id="my-video"></video>
```

I wonder if we can leverage the same inverted relationship where the button can query state about its linked idref?

```html
<style>
  button:invoketarget-linked(video:playing) .play-label,
  button:invoketarget-linked(video:paused) .pause-label {
    display:none
  }
</style>
<button id="A" invoketarget="my-video" invokeaction="playpause">
  <span class="play-label">Play</span>
  <span class="pause-label">Pause</span>
</button>
<video id="my-video"></video>
```

**Comment by lukewarlow** [comment]
Yeah Tim mentioned an ':invoketarget()' selector as an idea. Invalidating could be interesting but no more difficult than has or others.

**Comment by o-t-w** [comment]
I want to echo hidde's comment. 
It is now the case that the `:open` pseudo-class can be used to style any input with an open picker. e.g.

```css
input:open {
  border: solid 4px purple;
}
```

A difference is that the pseudo-class is on the input rather than a button. However, because the the only part of an input you can actually style is the button and not the picker, this feels like an almost identical use-case. It would give developers less stuff to learn.

**Comment by lukewarlow** [comment]
For the aria-expanded case it does seem like :open might make sense? Especially given selects get it and they will have a stylable select soon?

Might be weird that the button matches :open and the target matches :open (for dialog and openable) or :popover-open for popovers.

You also have the slight issue that :not(open) matches two states. A command button with a closed target AND a command button with an invalid (or non expandy) target. But maybe that's fine?

Though I do think we maybe need a wider discussion about CSS selectors to match aria state more widely.

**Comment by o-t-w** [comment]
> Though I do think we maybe need a wider discussion about CSS selectors to match aria state more widely.

Is a discussion about that happening? Where would be the best venue for that? Have you written anything about it?

**Comment by g12n** [comment]
I think we have to be precise here, what we actually want to select.

Lets start with a simple demo case

![Image](https://github.com/user-attachments/assets/aa98aaf6-301b-4fdb-888b-1ce7962fd2da)

```html
<div aria-label="products on this photo">
     <button popovertoggletarget="product-1">Book</button>
     <button commandfor ="product-2" command="toggle-popover">Mug</button>
</div>

<div popover id="product-1"><p>Pen<p><p>ab 15,99€</p></div>
<div popover id="product-2"><p>Pen<p><p>ab 15,99€</p></div>
``` 

## **Option 1:** The element with a `for` or `popovertoggletarget` or `commandfor` attribute based on the state of its targeted element.

This would be a pseudo class checking similar to `parent:has(child)` but for the invoker-target relationship. 

```css
button::has-target(:popover-open){
    --circle-size: 80%;
}
```
This would target any button with linked to an open popover. 

![Image](https://github.com/user-attachments/assets/f21afe6e-309e-45e5-b581-a0f6be2fe84d)

If there were two buttons invoking the same target, both would be selected. This would be similar to the [above mentioned](https://github.com/openui/open-ui/issues/622#issuecomment-1310824380) `body:has([id=pop]:open) [popovertoggletarget=pop] `

## **Option 2:** The element that invoked a given elements current state

This would be pseudo element selecting the element.

```css
:popover-open::invoker{
    --circle-size: 80%;
}
```
This would work similar to [implicit anchor implicit_anchor_association](https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Anchor_positioning/Using#implicit_anchor_association) and only target the element that has initiated the current state. It would select the most recent [CommandEvent: source](https://developer.mozilla.org/en-US/docs/Web/API/CommandEvent/source) or the [ToggleEvent:source](https://developer.mozilla.org/en-US/docs/Web/API/ToggleEvent/source).

![Image](https://github.com/user-attachments/assets/41853e83-5153-4d8d-871c-98401120d1d5)

This might even be more implicit by additional selectors `::invoker(button)`.
